core/src/test/java/org/infinispan/persistence/sifs/SoftIndexFileStoreFileStatsTest.java

```diff
@@ -269,4 +271,48 @@ public void testExpirationStats(Method m, boolean extraRemovedEntry) throws Inte
          TestingUtil.replaceComponent(cacheManager, TimeService.class, controlledTimeService.getActualTimeService(), true);
       }
    }
+
+   public void testExpirationCompactionOnLogFile(Method m) throws InterruptedException {
+      String cacheName = m.getName();
+
+      ControlledTimeService controlledTimeService = defineCacheConfigurationAndInjectTimeService(cacheName);
+      try {
+         Cache<String, Object> cache = cacheManager.getCache(cacheName);
+         cache.start();
+
+         cache.put("expired", "bar", 10, TimeUnit.SECONDS);
+
+         // This value will be overwritten in a loop later to cause log file to roll over
+         cache.put("replace-me", "1");
+
+         controlledTimeService.advance(TimeUnit.SECONDS.toMillis(11));
+
+         NonBlockingSoftIndexFileStore store = TestingUtil.getFirstStore(cache);
+
+         Compactor compactor = TestingUtil.extractField(store, "compactor");
+
+         // We use a sync queue so we can block the compaction and insert values
+         SynchronousQueue<Object> queue = new SynchronousQueue<>();
+
+         MyCompactionObserver myCompactionObserver = new MyCompactionObserver(queue);
+         // This will be blocked by the expired entry above
+         compactor.performExpirationCompaction(myCompactionObserver);
+
+         int increment = 2;
+         ConcurrentMap<Integer, Compactor.Stats> fileStats = compactor.getFileStats();
+         while (extractCompletedStat(fileStats) == null) {
+            cache.put("replace-me", String.valueOf(increment++));
+         }
+
+         // Let compaction expiration complete now
+         assertNotNull(queue.poll(10, TimeUnit.SECONDS));
+
+         // Compactor should clean up the old file and it should just be a single entry for the removed entry
+         eventually(() -> "File stats are: " + fileStats + " and data directory size is: " +
+                     SoftIndexFileStoreTestUtils.dataDirectorySize(tmpDirectory, cacheName)
+               , () -> SoftIndexFileStoreTestUtils.dataDirectorySize(tmpDirectory, cacheName) < 1000L);
```

Comment:
Maybe we do a check on the size before and after compaction?

Reply:
The compaction is triggered from the previous one, but isn't controlled by us. So no way to really do that. It is done asynchronously from our calls, so the eventually should be fine imo.

Reply:
If you are curious if the data directory was always less than 1000, it can't be possible as the max file size is 1000 and the lines above force it to have at least that much data.